modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -505,6 +543,148 @@ public int getZoneId(String name) {
+     * based on the topology with passed or greater version. Since the date nodes value is updated asynchronously, this method waits for
+     * the date nodes to be updated with new nodes in the topology if the value of auto adjust scale up is 0. And also waits for
+     * the date nodes to be updated with nodes that have left the topology if the value of auto adjust scale down is 0.
+     * After the zone manager has observed the logical topology change and the data nodes value is updated according to cluster topology,
+     * then this method completes the returned future with the current value of data nodes.
+     *
+     * <p>If the value of auto adjust scale up is greater than zero, then it is not necessary to wait for the data nodes update triggered
+     * by new nodes in cluster topology. Similarly if the value of auto adjust scale down is greater than zero, then it is not necessary to
+     * wait for the data nodes update triggered by new nodes that have left the topology in cluster topology.
+     *
+     * <p>The returned future can be completed with {@link DistributionZoneNotFoundException} and
+     * {@link DistributionZoneWasRemovedException} in case when the distribution zone was removed during method execution.
+     *
+     * @param zoneId Zone id.
+     * @param topVer Topology version.
+     * @return The data nodes future which will be completed with data nodes for the zoneId or with exception.
+     */
+    public CompletableFuture<Set<String>> topologyVersionedDataNodes(int zoneId, long topVer) {
+        CompletableFuture<IgniteBiTuple<Boolean, Boolean>> timerValuesFut = awaitTopologyVersion(topVer)
+                .thenCompose(ignored -> getImmediateTimers(zoneId));
+
+        return allOf(
+                timerValuesFut.thenCompose(timerValues -> scaleUpAwaiting(zoneId, timerValues.get1())),
+                timerValuesFut.thenCompose(timerValues -> scaleDownAwaiting(zoneId, timerValues.get2()))
+        ).thenCompose(ignored -> getDataNodesFuture(zoneId));
+    }
+
+    /**
+     * Waits for DistributionZoneManager waits for observing passed topology version or greater version
+     * in {@link DistributionZoneManager#topologyWatchListener}.
+     *
+     * @param topVer Topology version.
+     * @return Future for chaining.
+     */
+    private CompletableFuture<Void> awaitTopologyVersion(long topVer) {
+        return inBusyLock(busyLock, () -> topVerTracker.waitFor(topVer));
+    }
+
+    /**
+     * Transforms {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleUp}
+     * and {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleDown} values to boolean values.
+     * True if it equals to zero and false if it greater than zero. Zero means that data nodes changing must be started immediately.
+     *
+     * <p>The returned future can be completed with {@link DistributionZoneNotFoundException}
```

Comment:
It's kinda strange that it completes with `DistributionZoneNotFoundException` in case when the distribution zone was removed, but not with `DistributionZoneWasRemovedException`

Reply:
I've updated the java doc.